echopype/clean/impulse_noise.py

```diff
@@ -0,0 +1,463 @@
+"""
+Algorithms for masking Impulse noise.
+
+    Copyright (c) 2020 Echopy
+
+    This code is licensed under the MIT License.
+    See https://opensource.org/license/mit/ for details.
+    Original code sourced from:
+    https://github.com/open-ocean-sounding/echopy/blob/master/echopy/processing/mask_impulse.py
+
+"""
+__authors__ = [
+    "Alejandro Ariza",  # wrote ryan(), ryan_iterable(), and wang()
+    "Raluca Simedroni",  # adapted the impulse noise masking algorithms to echopype
+]
+
+import warnings
+from typing import List, Tuple, Union
+
+import numpy as np
+import xarray as xr
+from scipy.ndimage import median_filter as medianf
+from skimage.morphology import dilation, erosion
+
+from ..utils import mask_transformation
+
+
+def _ryan(
+    Sv: np.ndarray, iax: np.ndarray, m: Union[int, float], n: int, thr: Union[int, float]
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Mask impulse noise following the two-sided comparison method described
+    in:
+        Ryan et al. (2015) ‘Reducing bias due to noise and attenuation in
+        open-ocean echo integration data’, ICES Journal of Marine Science,
+        72: 2482–2493.
+
+    Parameters
+    ----------
+        Sv  (float)    : 2D array with Sv data to be masked (dB).
+        iax (int/float): 1D array with i axis data (n samples or range).
+        m   (int/float): vertical binning length (n samples or range).
+        n   (int)      : number of pings either side for comparisons.
+        thr (int/float): user-defined threshold value (dB).
+
+    Returns
+    -------
+        bool: 2D array with IN mask
+        bool: 2D array with mask indicating valid IN mask samples.
+    """
+
+    # resample down vertically
+    iax_ = np.arange(iax[0], iax[-1], m)
+    Sv_ = mask_transformation.oned(Sv, iax, iax_, 0, log_var=True)[0]
+
+    # resample back to full resolution
+    jax = np.arange(len(Sv[0]))
+    Sv_, mask_ = mask_transformation.full(Sv_, iax_, jax, iax, jax)
+
+    # side comparison (±n)
+    dummy = np.zeros((iax.shape[0], n)) * np.nan
+    comparison_forward = Sv_ - np.c_[Sv_[:, n:], dummy]
+    comparison_backward = Sv_ - np.c_[dummy, Sv_[:, 0:-n]]
```

Comment:
[xr.DataArray.shift](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.shift.html#xarray.DataArray.shift) likely useful here for maintaining things as `xr.DataArrays`.